crates/fluvio/src/producer/accumulator.rs

```diff
@@ -0,0 +1,167 @@
+    batch_size: usize,
+    batches: Arc<HashMap<PartitionId, Arc<Mutex<VecDeque<ProducerBatch>>>>>,
+}
+
+impl RecordAccumulator {
+    pub(crate) fn new(batch_size: usize, partition_n: i32) -> Self {
+        let mut batches = HashMap::default();
+        for i in 0..partition_n {
+            batches.insert(i, Arc::new(Mutex::new(VecDeque::new())));
+        }
+        Self {
+            batches: Arc::new(batches),
+            batch_size,
+        }
+    }
+
+    pub(crate) async fn push_record(
+        &self,
+        record: Record,
+        partition_id: PartitionId,
+    ) -> Result<PushRecord> {
+        let batches_lock = self
+            .batches
+            .get(&partition_id)
+            .ok_or(ProducerError::PartitionNotFound(partition_id))?;
+
+        let mut batches = batches_lock.lock().await;
+        if let Some(batch) = batches.back_mut() {
+            if let Ok(push_record) = batch.push_record(record.clone()) {
+                return Ok(PushRecord::new(push_record, batch.is_full(), false));
+            }
+        }
+
+        let mut batch = ProducerBatch::new(self.batch_size);
+
+        match batch.push_record(record) {
+            Ok(push_record) => {
+                let batch_is_full = batch.is_full();
+
+                batches.push_back(batch);
+
+                Ok(PushRecord::new(push_record, batch_is_full, true))
+            }
+            Err(err) => Err(err),
+        }
+    }
+
+    pub(crate) fn batches(&self) -> Arc<HashMap<PartitionId, Arc<Mutex<VecDeque<ProducerBatch>>>>> {
+        self.batches.clone()
+    }
+}
+
+pub(crate) struct PushRecord {
+    pub(crate) future: FutureRecordMetadata,
+    pub(crate) is_full: bool,
+    pub(crate) new_batch: bool,
+}
+
+impl PushRecord {
+    fn new(future: FutureRecordMetadata, is_full: bool, new_batch: bool) -> Self
+where {
+        Self {
+            future,
+            is_full,
+            new_batch,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct ProducerBatch {
+    write_limit: usize,
+    current_size: usize,
+    is_full: bool,
+    create_time: Instant,
+    pub(crate) records: Vec<AccumulatedRecord>,
+}
+impl ProducerBatch {
+    fn new(write_limit: usize) -> Self {
+        let now = Instant::now();
+
+        Self {
+            is_full: false,
+            write_limit,
+            create_time: now,
+            current_size: 0,
+            records: vec![],
+        }
+    }
+
+    pub(crate) fn create_time(&self) -> &Instant {
+        &self.create_time
+    }
+
+    fn push_record(&mut self, record: Record) -> Result<FutureRecordMetadata> {
+        let relative_offset = self.records.len() as i64;
+        let (sender, receiver) = async_channel::bounded(1);
+        let record_size = record.write_size(0);
+
+        if self.current_size + record_size > self.write_limit {
+            self.is_full = true;
+            return Err(FluvioError::Other("cannot be added to this batch".into()));
```

Comment:
We cannot add that to this batch because it doesnt have enough space, when this error happens, we are creating a new batch and adding the records there, see line 52

Reply:
Then why need to return error here?